usersession/client/client.go

```diff
@@ -0,0 +1,268 @@
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) doMany(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
+	sockets, err := filepath.Glob(filepath.Join(dirs.XdgRuntimeDirGlob, "snapd-session-agent.socket"))
+	if err != nil {
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				// Ignore directories that do not
+				// appear to be valid XDG runtime dirs
+				// (i.e. /run/user/NNNN).
+				return
+			}
+			response := response{uid: uid}
+			defer func() {
+				mu.Lock()
+				defer mu.Unlock()
+				responses = append(responses, &response)
+			}()
+
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				response.err = err
+				return
+			}
+			req = req.WithContext(ctx)
+			for key, value := range headers {
+				req.Header.Set(key, value)
+			}
+			httpResp, err := client.doer.Do(req)
+			if err != nil {
+				response.err = err
+				return
+			}
+			defer httpResp.Body.Close()
+			response.statusCode = httpResp.StatusCode
+			response.err = decodeInto(httpResp.Body, &response)
+			response.checkError()
+		}(socket)
+	}
+	wg.Wait()
+	return responses, nil
+}
+
+func decodeInto(reader io.Reader, v interface{}) error {
+	dec := json.NewDecoder(reader)
+	if err := dec.Decode(v); err != nil {
+		r := dec.Buffered()
+		buf, err1 := ioutil.ReadAll(r)
+		if err1 != nil {
+			buf = []byte(fmt.Sprintf("error reading buffered response body: %s", err1))
+		}
+		return fmt.Errorf("cannot decode %q: %s", buf, err)
+	}
+	return nil
+}
+
+type SessionInfo struct {
+	Version string `json:"version"`
+}
+
+func (client *Client) SessionInfo(ctx context.Context) (info map[int]SessionInfo, err error) {
+	responses, err := client.doMany(ctx, "GET", "/v1/session-info", nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	info = make(map[int]SessionInfo)
+	for _, resp := range responses {
+		if resp.err != nil {
+			if err == nil {
+				err = resp.err
+			}
+			continue
+		}
+		var si SessionInfo
+		if decodeErr := json.Unmarshal(resp.Result, &si); decodeErr != nil {
+			if err == nil {
+				err = decodeErr
+			}
+			continue
+		}
+		info[resp.uid] = si
+	}
+	return info, err
+}
+
+type ServiceFailure struct {
+	Uid     int
+	Service string
+	Error   string
+}
+
+func decodeServiceErrors(uid int, errorValue map[string]interface{}, kind string) []ServiceFailure {
```

Comment:
this should return an error if needed

Reply:
I've reorganised this function to add an error return.  Note that in this code path we've also got an error message from the agent in the body.  At present, the Services*() method return a single error, so I thought it would be better to return that error to the caller.

I guess the alternative would be to switch over to reporting a list of errors rather than the first, but that kind of thing doesn't seem to be common elsewhere in the code base.